Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp

```diff
@@ -60,30 +75,92 @@ std::unique_ptr<DisplayList::DisplayList> SkiaThreadedPaintingPool::recordDispla
     return displayList;
 }
 
-void SkiaThreadedPaintingPool::postPaintingTask(Ref<Nicosia::Buffer>& buffer, const TiledBackingStore& tiledBackingStore, const CoordinatedGraphicsLayer& layer, const IntRect& dirtyRect)
+bool SkiaThreadedPaintingPool::paintDisplayListIntoBuffer(Ref<Nicosia::Buffer>& buffer, DisplayList::DisplayList& displayList)
+{
+    auto* canvas = buffer->canvas();
+    if (!canvas)
+        return false;
+
+    static thread_local RefPtr<ControlFactory> s_controlFactory;
+    if (!s_controlFactory)
+        s_controlFactory = ControlFactory::create();
+
+    canvas->save();
+    canvas->clear(SkColors::kTransparent);
+
+    GraphicsContextSkia context(*canvas, buffer->isBackedByOpenGL() ? RenderingMode::Accelerated : RenderingMode::Unaccelerated, RenderingPurpose::LayerBacking);
+    context.drawDisplayListItems(displayList.items(), displayList.resourceHeap(), *s_controlFactory, FloatPoint::zero());
+
+    canvas->restore();
+    return true;
+}
+
+bool SkiaThreadedPaintingPool::paintGraphicsLayerIntoBuffer(Ref<Nicosia::Buffer>& buffer, const TiledBackingStore& tiledBackingStore, const CoordinatedGraphicsLayer& layer, const IntRect& dirtyRect) const
+{
+    auto* canvas = buffer->canvas();
+    if (!canvas)
+        return false;
+
+    canvas->save();
+    canvas->clear(SkColors::kTransparent);
+
+    GraphicsContextSkia context(*canvas, buffer->isBackedByOpenGL() ? RenderingMode::Accelerated : RenderingMode::Unaccelerated, RenderingPurpose::LayerBacking);
+    layer.paintIntoGraphicsContext(context, tiledBackingStore, dirtyRect);
+
+    canvas->restore();
+    return true;
+}
+
+bool SkiaThreadedPaintingPool::canPerformAcceleratedRendering() const
+{
+    return ProcessCapabilities::canUseAcceleratedBuffers() && PlatformDisplay::sharedDisplay().skiaGLContext()->makeContextCurrent();
+}
```

Comment:
This has the side effect of making skia context current, I think it would be better to check if there's a skia gl context created.

Reply:
Sounds reasonable, will adapt.

Reply:
Oh I remember now: I want to fallback to CPU rendering, if makeContextCurrent() actually fails.
Since that function is used to decide whether to use GPU rendering or not, kaing the context current has to be part of this function -- it's on purpose.